File: internal/update/github.go
package update

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"golang.org/x/mod/semver"
)

const (
	defaultOwner = "brandonhsz"
	defaultRepo  = "TopoDuctor"
)

// ReleaseTag names the latest published release tag (e.g. "v1.2.3").
type ReleaseTag struct {
	Tag string
	URL string
}

type githubRelease struct {
	TagName string `json:"tag_name"`
	HTMLURL string `json:"html_url"`
}

// FetchLatestRelease queries GitHub's latest release API.
func FetchLatestRelease(ctx context.Context, owner, repo string) (ReleaseTag, error) {
	if owner == "" {
		owner = defaultOwner
	}
	if repo == "" {
		repo = defaultRepo
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet,
		fmt.Sprintf("https://api.github.com/repos/%s/%s/releases/latest", owner, repo), nil)
	if err != nil {
		return ReleaseTag{}, err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "topoductor-update-check")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return ReleaseTag{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return ReleaseTag{}, fmt.Errorf("GitHub API: %s", resp.Status)
	}

	var body githubRelease
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return ReleaseTag{}, err
	}
	tag := strings.TrimSpace(body.TagName)
	if tag == "" {
		return ReleaseTag{}, fmt.Errorf("respuesta sin tag_name")
	}
	return ReleaseTag{Tag: tag, URL: body.HTMLURL}, nil
}

// Normalize semver for golang.org/x/mod/semver (must start with "v").
func Normalize(v string) string {
	v = strings.TrimSpace(v)
	v = strings.TrimPrefix(v, "v")
	if v == "" || strings.EqualFold(v, "dev") {
		return "v0.0.0"
	}
	i := strings.IndexByte(v, '-')
	if i >= 0 {
		v = v[:i]
	}
	if !semver.IsValid("v" + v) {
		return "v0.0.0"
	}
	return "v" + v
}

// IsNewer reports true if latest is strictly greater than current (both semver-normalized).
func IsNewer(current, latest string) bool {
	c := Normalize(current)
	l := Normalize(latest)
	return semver.Compare(c, l) < 0
}

File: src/update/github.ts
import semver from "semver";

const defaultOwner = "brandonhsz";
const defaultRepo = "TopoDuctor";

export type ReleaseTag = { tag: string; url: string };

type GithubRelease = {
  tag_name: string;
  html_url: string;
};

export async function fetchLatestRelease(
  owner = defaultOwner,
  repo = defaultRepo,
  signal?: AbortSignal
): Promise<ReleaseTag> {
  const url = `https://api.github.com/repos/${owner}/${repo}/releases/latest`;
  const res = await fetch(url, {
    signal,
    headers: {
      Accept: "application/vnd.github+json",
      "User-Agent": "topoductor-update-check",
    },
  });
  if (!res.ok) {
    throw new Error(`GitHub API: ${res.status} ${res.statusText}`);
  }
  const body = (await res.json()) as GithubRelease;
  const tag = (body.tag_name ?? "").trim();
  if (!tag) {
    throw new Error("respuesta sin tag_name");
  }
  return { tag, url: body.html_url };
}

export function normalizeSemver(v: string): string {
  v = v.trim().replace(/^v/i, "");
  if (!v || v.toLowerCase() === "dev") {
    return "v0.0.0";
  }
  const i = v.indexOf("-");
  if (i >= 0) {
    v = v.slice(0, i);
  }
  const withV = v.startsWith("v") ? v : `v${v}`;
  if (!semver.valid(withV)) {
    return "v0.0.0";
  }
  return withV;
}

export function isNewerThan(current: string, latest: string): boolean {
  return semver.lt(normalizeSemver(current), normalizeSemver(latest));
}
